### PR TITLE
Feature/result proposal obi

### PR DESCRIPTION
**Pull request by ObiWahn**

### CODE REVIEW

**Comment by danielhlarkin** [comment]
So to be clear, are we explicitly disallowing reference types? If not, we need a solution to the following:
```
/home/dan/work/ArangoDevel/lib/Basics/Result.h:133:3: note: ‘arangodb::TypedResult<T>::TypedResult() [with T = int&]’ is implicitly deleted because the default definition would be ill-formed:
   TypedResult() = default;
   ^~~~~~~~~~~
```

**Comment by ObiWahn** [comment]
Hey Dan,

I have talked with @jsteemann this morning and he is happy to drop support for references in Results so we pass real values around. I have still tried to fix the ref support a bit in case we consider keeping it. I have added a test file that we could use in the future to document our requirements. I wonder if you tried to default construct something like `TypedResult<int&>`! What kind of code created the above error?

**Comment by ObiWahn** [comment]
I think there is still a problem with rvalue refs with types that are not move constructible.

**Comment by danielhlarkin** [comment]
@ObiWahn Yes, that's exactly what I did. `TypedResult<int&> intRefResult{};`.

I guess we should probably explicitly adopt the convention that we return a reference (or other raw type, no result) if the operation is guaranteed to succeed. If it is not guaranteed to succeed, we should return a `Result` or `TypedResult`, and use actual values or pointers. 

Could you give an example of the rvalue issue you mentioned?

**Comment by ObiWahn** [comment]
@dan a ref must be initialized at creation time. `int& foo;` will not work either.

**Comment by danielhlarkin** [comment]
@ObiWahn Yes. I just wanted to clarify how we were treating refs. Seems either we disable any non-value constructor for ref types, or we disallow ref types altogether. Either is fine. I would lean toward the latter since it will likely make things a bit simpler. (Also see above about the guarantee of success.)

File: lib/Basics/Result.h [hunk 1]
```diff
@@ -18,89 +18,112 @@
...
+/// @author Jan Christoph Uhde
 ////////////////////////////////////////////////////////////////////////////////
 
 #ifndef ARANGODB_BASICS_RESULT_H
 #define ARANGODB_BASICS_RESULT_H 1
 
 #include "Basics/Common.h"
+#include "Basics/error.h"
+#include <type_traits>
+#include <iostream>
```
Comment (jsteemann) [inline lib/Basics/Result.h]: including iostream in a header file that is included about everywhere is not a good idea.
is it actually necessary at all? if yes, can we get away with `#include <iosfws>`?

File: lib/Basics/Result.h [hunk 2]
```diff
@@ -18,89 +18,112 @@
...
-    }
-  }
+  Result(int errorNumber) noexcept
+    : _errorNumber(errorNumber)
+    {}
 
   Result(int errorNumber, std::string const& errorMessage)
-      : _errorNumber(errorNumber), _errorMessage(errorMessage) {}
+    : _errorNumber(errorNumber)
+    , _errorMessage(std::make_unique<std::string>(errorMessage))
```
Comment (jsteemann) [inline lib/Basics/Result.h]: +1 allocation compared to previous Result implementation. Not good.

File: lib/Basics/Result.h [hunk 3]
```diff
@@ -18,89 +18,112 @@
...
+    {}
 
   Result(int errorNumber, std::string&& errorMessage)
-      : _errorNumber(errorNumber), _errorMessage(std::move(errorMessage)) {}
-  
+    : _errorNumber(errorNumber)
+    , _errorMessage(std::make_unique<std::string>(std::move(errorMessage)))
```
Comment (jsteemann) [inline lib/Basics/Result.h]: +1 allocation compared to previous Result implementation. Not good.

File: lib/Basics/Result.h [hunk 4]
```diff
@@ -18,89 +18,112 @@
...
+    {}
+
   // copy
-  Result(Result const& other) 
-      : _errorNumber(other._errorNumber), 
-        _errorMessage(other._errorMessage) {}
+  Result(Result const& other)
+    : _errorNumber(other._errorNumber)
+    , _errorMessage( other._errorMessage
+                   ? std::make_unique<std::string>(*other._errorMessage)
```
Comment (jsteemann) [inline lib/Basics/Result.h]: +1 allocation compared to previous Result implementation. Not good.

File: lib/Basics/Result.h [hunk 5]
```diff
@@ -18,89 +18,112 @@
...
-  virtual ~Result() {}
-
  public:
+  int errorNumber() const { return _errorNumber; }
+  std::string errorMessage() const {
+    if (!_errorMessage) {
+      if(_errorNumber != TRI_ERROR_NO_ERROR){
+        _errorMessage = std::make_unique<std::string>(TRI_errno_string(_errorNumber));
+      } else {
+        _errorMessage = std::make_unique<std::string>();
```
Comment (jsteemann) [inline lib/Basics/Result.h]: why not simply `return std::string();`? would save one allocation

File: lib/Basics/Result.h [hunk 6]
```diff
@@ -18,89 +18,112 @@
...
     _errorNumber = errorNumber;
-    _errorMessage = errorMessage;
+    _errorMessage = std::make_unique<std::string>(errorMessage);
     return *this;
   }
 
   Result& reset(int errorNumber, std::string&& errorMessage) noexcept {
     _errorNumber = errorNumber;
-    _errorMessage = std::move(errorMessage);
+    _errorMessage = std::make_unique<std::string>(std::move(errorMessage));
```
Comment (jsteemann) [inline lib/Basics/Result.h]: +1 allocation compared to previous Result implementation. Not good.

File: lib/Basics/Result.h [hunk 7]
```diff
@@ -18,89 +18,112 @@
...
     return *this;
   }
 
   Result& reset(Result const& other) {
     _errorNumber = other._errorNumber;
-    _errorMessage = other._errorMessage;
+    _errorMessage = other._errorMessage
+                  ? std::make_unique<std::string>(*other._errorMessage)
```
Comment (jsteemann) [inline lib/Basics/Result.h]: +1 allocation compared to previous Result implementation. Not good.

File: lib/Basics/Result.h [hunk 8]
```diff
@@ -110,16 +133,104 @@ class Result {
...
-  // the default implementation is const, but sub-classes might
-  // really do more work to compute.
-
-  virtual std::string errorMessage() const& { return _errorMessage; }
-  virtual std::string errorMessage() && { return std::move(_errorMessage); }
-
  protected:
   int _errorNumber;
-  std::string _errorMessage;
+  mutable std::unique_ptr<std::string> _errorMessage;
```
Comment (jsteemann) [inline lib/Basics/Result.h]: why does this have to be mutable?

File: lib/Basics/Result.h [hunk 9]
```diff
@@ -110,16 +133,104 @@ class Result {
...
+           >
+  TypedResult(ValueType value
+             ,int error = TRI_ERROR_NO_ERROR
+             ,std::string message = ""
+             )
+    : value(value)
+    , _valid(true)
+    ,_result(error, std::move(message))
+    {
+      std::cerr << "copy" << std::endl;
```
Comment (jsteemann) [inline lib/Basics/Result.h]: Is this output supposed to be there?

File: lib/Basics/Result.h [hunk 10]
```diff
@@ -110,16 +133,104 @@ class Result {
...
+           >
+  TypedResult(ValueType&& value
+             ,int error = TRI_ERROR_NO_ERROR
+             ,std::string message = ""
+             )
+    : _valid(true)
+    ,_result(error, std::move(message))
+    {
+      value = std::move(value);
+      std::cerr << "move assign" << std::endl;
```
Comment (jsteemann) [inline lib/Basics/Result.h]: is this output supposed to be there?

File: lib/Basics/Result.h [hunk 11]
```diff
@@ -110,16 +133,104 @@ class Result {
...
+    ,_result(error, std::move(message))
+    {
+      std::cerr << "copy" << std::endl;
+    }
+
+  template <int x = !std::is_lvalue_reference<T>::value
+                    && std::is_move_constructible<T>::value
+           ,typename std::enable_if<x>::type* = nullptr
+           >
+  TypedResult(ValueType value
```
Comment (jsteemann) [inline lib/Basics/Result.h]: why exactly these overloads for `bool`, `int` and `uint32_t`?
Are these just the initial few types we wanted to overload? Do we need to add all other types here?

File: tests/Basics/ResultTest.cpp [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    auto rv = function_a(42); // create one result and try modify / reuse
+                              // it to make copy elision happen
+
+    if(rv.ok()) {
+      CHECK(rv.value == 42); // do something with the value
+    } else {
+      rv.reset(rv.errorNumber(), "error in function_b: " + rv.errorMessage());
+    }
+
+    if(false){
```
Comment (jsteemann) [inline tests/Basics/ResultTest.cpp]: intention of this?

File: tests/Basics/ResultTest.cpp [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+
+TEST_CASE("ResultTest", "[string]") {
+
+////////////////////////////////////////////////////////////////////////////////
+/// @brief test_StringBuffer1
+////////////////////////////////////////////////////////////////////////////////
+
+SECTION("test_ResultTest1") {
+  int  integer = 43;
+  int& integer_lvalue_ref = integer;
```
Comment (jsteemann) [inline tests/Basics/ResultTest.cpp]: coding conventions for variable names not followed (ff)

File: tests/Basics/ResultTest.cpp [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+  int const&  integer_lvalue_cref = integer;
+  int* integer_ptr = &integer;
+  std::string str = "arangodb rocks";
+
+  struct no_move {
```
Comment (jsteemann) [inline tests/Basics/ResultTest.cpp]: coding conventions for class names not followed.

**Comment by danielhlarkin** [comment]
@jsteemann One of the main changes to `Result` itself is to use lazy error message generation. It is a very common use case to return a `Result` rather than a raw `int`, even when we don't really care about the error message. This is partially to keep return types uniform and partially to allow access to convenience methods like `ok()`.

In any case, doing lazy generation for error messages allows us to avoid *any* allocation until the `errorMessage()` function is called if the default error message for the error code is used. It also avoids the map lookup in `TRI_errno_string()`.

In the case that it involves an allocation, we are likely already going to do an allocation within `std::string` itself, since most error messages are longer than the internal small-value optimization allows.

So it's a trade-off. In the old version we consistently do a map lookup and one allocation. In the new version, we do 0 allocations and no map lookup in one case, and we do 2 allocations and a map lookup in the other.

It is unclear which usage of `Result` is more common and what effect this will have on performance, but I would hazard a guess that this improves things on the whole. Perhaps we could conduct some experiments, @ObiWahn?

**Pull request closed without merging**